plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +64,110 @@ becomes
+          args'
+      _ -> pure Nothing
+
+    -- Checks whether it is safe (i.e., preserves semantics) to perform beta reduction,
+    -- namely, turning `(\a -> body) arg` into `body [a := arg]`.
+    safeToSubst :: name -> Term tyname name uni fun ann -> InlineM tyname name uni fun ann Bool
+    safeToSubst n = effectSafe (varBody info) Strict n <=< checkPurity
+
+-- | Consider whether to inline an application.
+inlineSaturatedApp ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  Term tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+inlineSaturatedApp t
+  | (Var _ann name, args) <- splitApplication t =
+      gets (lookupVarInfo name) >>= \case
+        Just varInfo -> fullyApplyAndBetaReduce varInfo args >>= \case
+          Just fullyApplied -> do
+            let def = varDef varInfo
+                -- Inline only if the size is no bigger than not inlining.
+                sizeIsOk = termSize fullyApplied <= termSize t
                 -- The definition itself will be inlined, so we need to check that the cost
                 -- of that is acceptable. Note that we do _not_ check the cost of the _body_.
                 -- We would have paid that regardless.
-                -- Consider e.g. `let y = \x. f x`. We pay the cost of the `f x` at every call
-                -- site regardless. The work that is being duplicated is the work for the lambda.
-                defCostOk = costIsAcceptable def
+                -- Consider e.g. `let y = \x. f x`. We pay the cost of the `f x` at
+                -- every call site regardless. The work that is being duplicated is
+                -- the work for the lambda.
+                costIsOk = costIsAcceptable def
             -- check if binding is pure to avoid duplicated effects.
-            -- For strict bindings we can't accidentally make any effects happen less often than
-            -- it would have before, but we can make it happen more often.
+            -- For strict bindings we can't accidentally make any effects happen less often
+            -- than it would have before, but we can make it happen more often.
             -- We could potentially do this safely in non-conservative mode.
             defPure <- isTermBindingPure (varStrictness varInfo) def
-            if fullyApplied && bodySizeOk && defCostOk && defPure
-            then do
-                -- rename the term before substituting in
-                renamed <- renameTerm defAsInlineTerm
-                pure $ fillAppContext renamed ctx
-            else pure tm
-          -- The variable maybe a *recursive* let binding, in which case it won't be in the map,
-          -- and we don't process it. ATM recursive bindings aren't inlined.
-          Nothing -> pure tm
-      -- if the term being applied is not a `Var`, don't inline
-      _ -> pure tm
+            pure $ if sizeIsOk && costIsOk && defPure then fullyApplied else t
```

Comment:
If we are worried about the cost of `fullyApplyAndBetaReduce`, we can do the `costIsOk` and `defPure` checks first, these checks don't need anything from `fullyApplyAndBetaReduce`. If either of them are false, we stop. Otherwise, do the case match on `fullyApplyAndBetaReduce`.

Reply:
That shouldn't be a problem. `fullyApplyAndBetaReduce` is linear.